### PR TITLE
[LLM Runtime] Implement Streaming-LLM

### PR DESCRIPTION
**Pull request by DDEle**
## Type of Change: WIP

feature or bug fix or documentation or others
API changed or not

## Description

detail description 
JIRA ticket: xxx

## Expected Behavior & Potential Risk

the expected behavior that triggered by this PR 

## How has this PR been tested?

how to reproduce the test (including hardware information)

## Dependency Change?

any library dependency introduced or removed

### CODE REVIEW

File: intel_extension_for_transformers/llm/runtime/graph/models/opt/opt.cpp [hunk 1]
```diff
@@ -97,7 +104,7 @@ static bool opt_model_eval_internal(model_context& lctx, const model_token* toke
   */
   struct ne_tensor* position = d_ne_new_tensor_1d(ctx0, NE_TYPE_I32, N);
   for (int i = 0; i < N; ++i) {
-    ((int32_t*)position->data)[i] = n_past + i + OPT_POS_EMBD_OFFS;
+    ((int32_t*)position->data)[i] = n_total + i + OPT_POS_EMBD_OFFS;
```
Comment (DDEle) [inline intel_extension_for_transformers/llm/runtime/graph/models/opt/opt.cpp:107]: @zhentaoyu Am I right here?

Comment (zhentaoyu) [reply intel_extension_for_transformers/llm/runtime/graph/models/opt/opt.cpp:107]: logical pos is `n_past + i +offset`? `n_total` will be greater than `max_seq_length`?

File: intel_extension_for_transformers/llm/runtime/graph/models/opt/opt.cpp [hunk 2]
```diff
@@ -151,7 +158,7 @@ static bool opt_model_eval_internal(model_context& lctx, const model_token* toke
           2, 0, 3);
 
       // store transposed key and value to memory (k_v cache)
-      if (N >= 1) {
+      if (N >= 1) {  // TODO(Yi): why don't store to kv-cache in next-token inference?
```
[the remark this answers is not in the document]
Comment (zhentaoyu) [reply intel_extension_for_transformers/llm/runtime/graph/models/opt/opt.cpp:161]: Can you make your request more clear?

[the remark this answers is not in the document]
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/models/opt/opt.cpp:161]: why don't store to kv-cache in next-token inference? Other models don't have this extra if clouse.

File: intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_utils.h [hunk 3]
```diff
@@ -114,15 +114,17 @@ MODEL_API size_t model_set_state_data(struct model_context* ctx, uint8_t* src);
...
+                                       size_t n_token_capacity, size_t* n_token_total_out);
 MODEL_API bool model_save_session_file(struct model_context* ctx, const char* path_session, const model_token* tokens,
-                                       size_t n_token_count);
+                                       size_t n_token_total);
 
 // Run the model inference to obtain the logits and probabilities for the next
 // token. tokens + n_tokens is the provided batch of new tokens to process
-// n_past is the number of tokens to use from previous eval calls
+// n_past is the offset to which the kv is cached to
+// n_totle is the number of tokens evaluated in previous eval calls
```
[the remark this answers is not in the document]
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_utils.h]: Thanks! Fixed

File: intel_extension_for_transformers/llm/runtime/graph/models/opt/opt.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (a32543254) [reply intel_extension_for_transformers/llm/runtime/graph/models/opt/opt.cpp:162]: Not a problem, could be removed

**Comment by a32543254** [comment]
1. Alibi associate with index      bloom/mpt may need change alibi kernel
2. beam search                          may need change kv cache reorder part code
3. when model use gather to embed rathe then repo       opt/ starcoder may cannot infiinitely inference

**Comment by zhentaoyu** [comment]
some thoughts:
1. We may need to re-rope the recent remained tokens if we decide to discard some middle tokens (see the implementation of rope kernel):
image
the `m` is the pos idx of the token in the **cache** rather than the text
see the official implementation repo: https://github.com/mit-han-lab/streaming-llm/blob/main/streaming_llm/pos_shift/modify_llama.py#L101-L104C8
we can give the `shifted k/v pos tensor` to `rope` kernel like `llama.cpp`: https://github.com/ggerganov/llama.cpp/blob/c5650ed470f6264046431c8b5e43cbfe3e680d17/llama.cpp#L2726 
2. discard more middle tokens will accelerate speed, however, the acc loss is unknown to me (`llama.cpp` said not too much, see  below):
   
image
